File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/model/TimeRange.java
package de.westnordost.streetcomplete.quests.opening_hours.model;

import android.annotation.SuppressLint;

/** A time range from [start,end).
 */
public class TimeRange extends CircularSection
{
	public final boolean isOpenEnded;

	public TimeRange(int minutesStart, int minutesEnd)
	{
		this(minutesStart, minutesEnd, false);
	}

	public TimeRange(int minutesStart, int minutesEnd, boolean openEnded)
	{
		super(minutesStart, minutesEnd);
		isOpenEnded = openEnded;
	}

	@Override public boolean intersects(CircularSection o)
	{
		if(!(o instanceof TimeRange)) return false;
		TimeRange other = (TimeRange) o;

		if(isOpenEnded && other.getStart() >= getStart())
			return true;
		if(other.isOpenEnded && getStart() >= other.getStart())
			return true;
		if(loops() && other.loops())
			return true;
		if(loops() || other.loops())
			return other.getEnd() > getStart() || other.getStart() < getEnd();
		return other.getEnd() > getStart() && other.getStart() < getEnd();
	}

	@Override public boolean equals(Object other)
	{
		if(other == null || !(other instanceof TimeRange)) return false;
		TimeRange o = (TimeRange) other;
		return o.isOpenEnded == isOpenEnded && super.equals(o);
	}

	@Override public int hashCode()
	{
		return super.hashCode() * 2 + (isOpenEnded ? 1 : 0);
	}

	@Override public String toString()
	{
		return toStringUsing("-");
	}

	public String toStringUsing(String range)
	{
		StringBuilder sb = new StringBuilder();
		sb.append(timeOfDayToString(getStart()));
		int end = getEnd();
		if(end == 0) end = 60*24;
		if(getStart() != getEnd() || !isOpenEnded)
		{
			sb.append(range);
			sb.append(timeOfDayToString(end));
		}
		if(isOpenEnded)
		{
			sb.append("+");
		}
		return sb.toString();
	}

	@SuppressLint("DefaultLocale") private static String timeOfDayToString(int minutes)
	{
		return String.format("%02d:%02d", minutes / 60, minutes % 60);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/model/TimeRange.kt
package de.westnordost.streetcomplete.quests.opening_hours.model

import kotlinx.serialization.Serializable
import java.text.DateFormat
import java.time.*
import java.util.Locale

/** A time range from [start,end). The times are specified in minutes. */
@Serializable
data class TimeRange(val start: Int, val end: Int, val isOpenEnded: Boolean = false) : Comparable<TimeRange> {

    fun intersects(other: TimeRange): Boolean =
        isOpenEnded && other.start >= start ||
        other.isOpenEnded && start >= other.start ||
        loops && other.loops ||
        if (loops || other.loops) other.end > start || other.start < end
        else                      other.end > start && other.start < end

    val loops get() = end < start

    override fun compareTo(other: TimeRange): Int {
        // loopers come first,
        if (loops && !other.loops) return -1
        if (!loops && other.loops) return +1
        // then by start
        val result = start - other.start
        if (result != 0) return result
        // then by end
        return end - other.end
    }

    fun toStringUsing(locale: Locale, range: String): String {
        val sb = StringBuilder()
        sb.append(timeOfDayToString(locale, start))
        if (start != end || !isOpenEnded) {
            sb.append(range)
            var displayEnd = timeOfDayToString(locale, end)
            if (displayEnd == "00:00") displayEnd = "24:00"
            sb.append(displayEnd)
        }
        if (isOpenEnded) sb.append("+")
        return sb.toString()
    }

    override fun toString() = toStringUsing(Locale.GERMANY, "-")

    private fun timeOfDayToString(locale: Locale, minutes: Int): String {
        val seconds = (minutes % (24*60)) * 60L
        val todayAt = LocalDateTime.of(LocalDate.now(), LocalTime.ofSecondOfDay(seconds))
            .atZone(ZoneId.systemDefault())
            .toInstant()
            .toEpochMilli()
        return DateFormat.getTimeInstance(DateFormat.SHORT, locale).format(todayAt)
    }
}
